engine.go

```diff
@@ -582,17 +595,64 @@ func (t *T) Context() context.Context {
 	return ctx
 }
 
+// Cleanup registers a function to be called
+// when this property check is over.
+// Use it to clean up any resources acquired during the check.
+//
+// Cleanup functions are called in last-in, first-out order.
+//
+// Note that the context returned by [T.Context] is canceled
+// by the time Cleanup functions are run.
+func (t *T) Cleanup(f func()) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.cleanups = append(t.cleanups, f)
+}
+
 // cleanup runs any cleanup tasks associated with the property check.
 // It is safe to call multiple times.
 func (t *T) cleanup() {
-	t.mu.Lock()
-	defer t.mu.Unlock()
+	t.cleaning.Store(true)
+	defer t.cleaning.Store(false)
+
+	// If a cleanup function panics,
+	// we still want to run the remaining cleanup functions.
+	defer func() {
+		t.mu.Lock()
+		recurse := len(t.cleanups) > 0
+		t.mu.Unlock()
+
+		if recurse {
+			t.cleanup()
```

Comment:
Is this black magic taken from `testing`?

Reply:
Yeah, from here: https://cs.opensource.google/go/go/+/refs/tags/go1.24.0:src/testing/testing.go;l=1420-1427

(Minus the two recovery modes. We don't need divergent behavior in this case.)